### PR TITLE
RFC: Defer completion of download-firmware mailbox transaction until runtime FW

### PR DESCRIPTION
**Pull request by korran**
While writing tests, I noticed how tempting it is for the SoC, after sending the download-firmware mailbox command, if that command fails, check the CPTRA_FW_ERROR_NON_FATAL register to find out why. Unfortunately, this is a dangerous pattern as there is a race condition:  CPTRA_FW_ERROR_NON_FATAL is set N clock cycles after the mailbox transaction is aborted (by the transaction Drop handler), so it is possible that the SoC will read the error register too early and report a bogus error code.

To prevent SoC integrators from making this mistake, I suggest that we always set the error register before aborting the mailbox transaction. In addition, if there is some sort of fault (bus error, etc), we should also abort current and future mailbox transactions to prevent any SoC firmware getting stuck waiting for a mailbox response that will never come.

This PR

* Prevents the mailbox transaction from being dropped in fmc_alias.rs ([diff](https://github.com/chipsalliance/caliptra-sw/compare/main...korran:caliptra-sw:defer-mailbox-complete?expand=1#diff-346d800d15991ad85381d00bde37d578cff3084618db86cb6cf84f4172fc6dc3R121-R129))
* In ROM, FMC, and Runtime, ensures that any pending mailbox transactions are aborted as part of report_error. ([diff](https://github.com/chipsalliance/caliptra-sw/pull/185/commits/d9ece98b0bd047795fff1fa7e4151ab716eb9fa1#diff-53ff01fbd62e2739aa48e599d6fc1dd54997f6ef9917b3d756812f6245dc70a5R161-R168))
* Moves the completion of the download-firmware mailbox transaction from the ROM ([diff](https://github.com/chipsalliance/caliptra-sw/pull/185/commits/d9ece98b0bd047795fff1fa7e4151ab716eb9fa1#diff-346d800d15991ad85381d00bde37d578cff3084618db86cb6cf84f4172fc6dc3R78-R79)) to the runtime firmware ([diff](https://github.com/chipsalliance/caliptra-sw/pull/185/commits/d9ece98b0bd047795fff1fa7e4151ab716eb9fa1#diff-d3b60fade8088a1621f04d56f53ba3a1adf3abead5b70c9678de57bdf92b7411R70-R74)). Doing this is debatable; completing the mailbox transaction in the app fw prevents the SoC from using the mailbox for sha512 offload while Caliptra signs the FMC and app measurements).
* Fixes up a bunch of test firmware so it works with these changes.

Before:

```
ready_for_fw high after 463260 cycles
UART: [afmc] Waiting for Image .
<<< Executing mbox cmd 0x46574c44 (9164 bytes) from SoC
UART: [afmc] Received Image of size 9164 bytes
UART: [afmc] Image verified using Vendor ECC Key Index 0
UART: [afmc] Loading FMC at address 0x40000000 len 4752
UART: [afmc] Loading Runtime at address 0x40002000 len 3452
>>> mbox cmd response: success
UART: [afmc] Signing Cert with AUTHORITY.KEYID = 5
UART: [afmc] Erasing AUTHORITY.KEYID = 5
```

After:

```
ready_for_fw high after 463260 cycles
UART: [afmc] Waiting for Image .
<<< Executing mbox cmd 0x46574c44 (9080 bytes) from SoC
UART: [afmc] Received Image of size 9080 bytes
UART: [afmc] Image verified using Vendor ECC Key Index 0
UART: [afmc] Loading FMC at address 0x40000000 len 4808
UART: [afmc] Loading Runtime at address 0x40002000 len 3312
UART: [afmc] Signing Cert with AUTHORITY.KEYID = 5
UART: [afmc] Erasing AUTHORITY.KEYID = 5
UART: [afmc] PUB.X = F398BBA281062CAB73E54067E53F27734090FAD28B9E9C1FEB37D3D655BF99BE6D78B3B03BFBA3A49A92DF2CEC99C904
UART: [afmc] PUB.Y = 1698A61EAC916D3476226B72F4E933AB8EA1A39FBAF64AD66CD3FEDEEB241011A69F2F83081FF5613DB92D378D8F48AC
UART: [afmc] SIG.R = B95B70494589B80758A8BD3F7C3BB059363BB6311EF0812AB82A0E854C9C7011E875ADAA8ABADC65287E4E7C54799FFD
UART: [afmc] SIG.S = 626F7F044819ED7AFC8ECB6B856883D2788484769610F5DDA64F748D60357355053FBC77F59310DAF5E9D5729E445805
UART: [afmc] --
UART: [cold-reset] --
UART: [state] Locking Datavault
UART: [state] Locking ICCM
UART: [fht] Loading FHT @ 0x50002000
UART: [exit] Launching FMC @ 0x40000134
UART: 
UART: Running Caliptra FMC ...
UART: 
UART: [fmc] FHT Marker: 0x54484643
UART: [fmc] FHT Major Version: 0x0001
UART: [fmc] FHT Minor Version: 0x0000
UART: [fmc] FHT Manifest Addr: 0x50000000
UART: [fmc] FHT FMC CDI KV KeyID: 6
UART: [fmc] FHT FMC PrivKey KV KeyID: 7
UART: [fmc] FHT RT Load Address: 0x00000002
UART: [fmc] FHT RT Entry Point: 0x00000001
UART: [art] Extend PCRs
UART: [cold-reset] ++
UART: [cold-reset] --
UART: [exit] Launching RT @ 0x40002000
UART: 
UART:   ____      _ _       _               ____ _____
UART:  / ___|__ _| (_)_ __ | |_ _ __ __ _  |  _ \_   _|
UART: | |   / _` | | | '_ \| __| '__/ _` | | |_) || |
UART: | |__| (_| | | | |_) | |_| | | (_| | |  _ < | |
UART:  \____\__,_|_|_| .__/ \__|_|  \__,_| |_| \_\|_|
UART:                |_|
UART: 
UART: [rt] FHT Marker: 0x54484643
UART: [rt] FHT Major Version: 0x0001
UART: [rt] FHT Minor Version: 0x0000
UART: [rt] FHT Manifest Addr: 0x50000000
UART: [rt] FHT FMC CDI KV KeyID: 6
UART: [rt] FHT FMC PrivKey KV KeyID: 7
UART: [rt] FHT RT Load Address: 0x00000001
UART: [rt] FHT RT Entry Point: 0x00000001
>>> mbox cmd response: success
```

What do you think?

### CODE REVIEW

**Comment by korran** [comment]
One potential downside is that the SoC can't use the mailbox for sha512 while waiting for the ROM to sign the cert. If that's a problem, we could still do the report_error thing in this PR but report success in the old spot. If an error happens while signing, shoddy SoC fw will find out the next time they make a mailbox request, which will fail.

File: fmc/test-fw/test-rt/src/main.rs
Comment (rusty1968) [inline fmc/test-fw/test-rt/src/main.rs:125]: Hi, Kor. Is there a test case that triggers this failure mode, so we can compare the behavior with the fix and without the fix?

Comment (korran) [reply fmc/test-fw/test-rt/src/main.rs:125]: Yeah, pretty much any of @mhatrevi's failure test cases exercise this behavior:

https://github.com/chipsalliance/caliptra-sw/blob/main/rom/dev/tests/test_image_validation.rs#L141-L148

The test passes as-is, but after this PR is merged you could get rid of the wait-for-fatal error logic and just check the error-code directly:

https://github.com/chipsalliance/caliptra-sw/blob/main/rom/dev/tests/test_image_validation.rs#L141-L148

could become

```
assert_eq!(
        Err(ModelError::MailboxCmdFailed),
        hw.upload_firmware(&image_bundle.to_bytes().unwrap())
    );
assert_eq!(hw.soc_ifc().cptra_fw_error_non_fatal().read(), 
           OWNER_PUB_KEY_DIGEST_MISMATCH);
```

Comment (korran) [reply fmc/test-fw/test-rt/src/main.rs:125]: (In case it wasn't clear, removing this code will cause all of those failure tests to freeze as hw.upload_firmware() never gets a response).

Comment (rusty1968) [reply fmc/test-fw/test-rt/src/main.rs:125]: The error scenario  is clear.

**Review by rusty1968 — APPROVED** [review]

**Review by jhand2 — APPROVED** [review]

**Pull request closed without merging**
